firestore/src/include/firebase/firestore/set_options.h

```diff
@@ -143,6 +145,23 @@ class SetOptions final {
   std::vector<FieldPath> fields_;
 };
 
+/** Checks `lhs` and `rhs` for equality. */
+inline bool operator==(const SetOptions& lhs, const SetOptions& rhs) {
+  if (lhs.type_ != rhs.type_) {
+    return false;
+  }
+  std::unordered_set<FieldPath> lhs_field_set(lhs.fields_.begin(),
```

Comment:
I think a better (though more cumbersome) approach would be to use a `unordered_map` instead of the `vector` to store field paths. This will require changing a few call sites, but it looks like there aren't too many. Android uses a `HashSet` for its equivalent class (more specifically, on Android `SetOptions` contains a [`FieldMask`](https://github.com/firebase/firebase-android-sdk/blob/14935e9c9ac097379b49f7790aa19c15b35f2a7d/firebase-firestore/src/main/java/com/google/firebase/firestore/model/mutation/FieldMask.java#L33) which is a wrapper over `Set`) -- I don't recall why we chose vector originally, but I strongly suspect that this could have something to do with us thinking that STLPort doesn't support `unordered_map`.

I'd suggest leaving out `SetOptions` out of this PR and creating a dedicated PR for it following this approach. If there are any unforeseen issues with that, please let me know.

Reply:
Is there a reason we should use `unordered_map` over `unordered_set` to store the field paths? IIUC we don't want duplicate FieldPath objects in `fields_`.

Reply:
I've posted https://github.com/firebase/firebase-cpp-sdk/pull/606 for this.